apps/src/lib/client/tm_jsonrpc_client.rs

```diff
@@ -0,0 +1,256 @@
+#[cfg(not(feature = "ABCI"))]
+mod tm_jsonrpc {
+    use std::convert::TryFrom;
+    use std::fmt::{Display, Formatter};
+    use std::ops::{Deref, DerefMut};
+
+    use curl::easy::{Easy2, Handler, WriteError};
+    use serde::{Deserialize, Serialize};
+    use tendermint_config::net::Address as TendermintAddress;
+    use tendermint_rpc::query::Query;
+
+    use crate::client::tendermint_rpc_types::{
+        parse, Error, EventParams, EventReply, TxResponse,
+    };
+
+    pub struct JsonRpcAddress<'a> {
+        host: &'a str,
+        port: u16,
+    }
+
+    impl<'a> TryFrom<&'a TendermintAddress> for JsonRpcAddress<'a> {
+        type Error = Error;
+
+        fn try_from(value: &'a TendermintAddress) -> Result<Self, Self::Error> {
+            match value {
+                TendermintAddress::Tcp { host, port, .. } => Ok(Self {
+                    host: host.as_str(),
+                    port: *port,
+                }),
+                _ => Err(Error::Address(value.to_string())),
+            }
+        }
+    }
+
+    impl<'a> Display for JsonRpcAddress<'a> {
+        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+            write!(f, "{}:{}", self.host, self.port)
+        }
+    }
+
+    /// The body of a json rpc request
+    #[derive(Serialize)]
+    pub struct Request {
+        /// Method name
+        pub method: String,
+        /// parameters to give the method
+        params: EventParams,
+        /// ID of the request
+        id: u8,
+    }
+
+    impl From<EventParams> for Request {
+        fn from(params: EventParams) -> Self {
+            Request {
+                method: "events".into(),
+                params,
+                id: 1,
```

Comment:
I'm guessing this ID is not significant, right?

Reply:
Nope, it's just there to conform with jsonrpc.